unit/solvers/smt2_incremental/convert_expr_to_smt.cpp

```diff
@@ -416,3 +416,75 @@ TEST_CASE(
     REQUIRE_THROWS(convert_expr_to_smt(unary_minus_exprt{true_exprt{}}));
   }
 }
+
+SCENARIO(
+  "Bitwise \"AND\" expressions are converted to SMT terms",
+  "[core][smt2_incremental]")
+{
+  GIVEN("three integer bitvectors and their smt-term equivalents")
+  {
+    const smt_termt smt_term_one = smt_bit_vector_constant_termt{1, 8};
+    const smt_termt smt_term_three = smt_bit_vector_constant_termt{3, 8};
+    const smt_termt smt_term_five = smt_bit_vector_constant_termt{5, 8};
+
+    const auto one_bvint = from_integer({1}, signedbv_typet{8});
+    const auto three_bvint = from_integer({3}, signedbv_typet{8});
+    const auto five_bvint = from_integer({5}, signedbv_typet{8});
+
+    WHEN("a bitand_exprt with two of them as arguments is converted")
+    {
+      const auto constructed_term =
+        convert_expr_to_smt(bitand_exprt{one_bvint, three_bvint});
+
+      THEN(
+        "it should be equivalent to a \"bvand\" term with the operands "
+        "converted as well")
+      {
+        const auto expected_term =
+          smt_bit_vector_theoryt::make_and(smt_term_one, smt_term_three);
+
+        CHECK(constructed_term == expected_term);
+      }
+    }
+
+    // bitand_exprt derives from multiary exprt, so we need to be able to handle
+    // expressions with more than 2 operands.
+    WHEN("a ternary bitand_exprt gets connverted to smt terms")
+    {
+      // We need to jump through a bit of a hoop because bitand_exprt doesn't
+      // support direct construction with multiple operands - so we have to
+      // construct its parent class and downcast it.
+      const exprt::operandst and_operands{one_bvint, three_bvint, five_bvint};
+      const multi_ary_exprt first_step{
+        ID_bitand, and_operands, signedbv_typet{8}};
+      const auto bitand_expr = to_bitand_expr(first_step);
```

Comment:
Indeed, but to me it looks like it makes intent-more explicit. I'm content with taking a small performance penalty (by a number of upcasts/downcasts) in tests for better readability (which leads to better debug-ability).

Is this something that you insist changes?

Reply:
I see no performance issue here. The issue to me is readability. The readability is not good either with or without your superfluous cast. I have a mild preference for leaving it out, because I don't think it really helps. The ideal solution here is just to use the constructor of `and_exprt`, rather than `multi_ary_exprt`. `and_exprt` has constructors which support 3 or more operands, which are suitable for your test.

As an aside, I think the constructor of `multi_ary_exprt` you used should be `protected`, not `public`, because the `_id` parameter is an implementation detail of the sub class. Leaving it exposed actually supports constructing expressions with 3 or more operands which aren't derived from `multi_ary_exprt`. For example `div_exprt` derives from `binary_exprt`, but there are no checks which prevent `ID_div` being used to construct `multi_ary_exprt` until a downcast is attempted later.